File: 199-binary-tree-right-side-view/binary-tree-right-side-view.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
    // return bfsSolution(root);
    if (!root) return [];
    let level = 0;
    let res = [];
    // let hm = {};
    return dfsSolution2nd(root, level, res);
};



function dfsSolution(cn, level, res, hm) {
    function dfsSol(cn, level) {
        if (hm[level] === undefined) { //!hm[level]
            res.push(cn.val);
            hm[level] = cn.val; // hm[level] = true
        }
        if (cn.right) dfsSol(cn.right, level + 1);
        if (cn.left) dfsSol(cn.left, level + 1);
        return res;
    }
    return dfsSol(cn, level);
}


function bfsSolution(root) {
    if (!root) return [];
    let q = [root];
    let res = [];
    while (q.length) {
        let clLength = q.length;//current level length
        let cn;
        for (let i = 0; i < clLength; i++) {
            cn = q.shift();
            if (cn.left) q.push(cn.left);
            if (cn.right) q.push(cn.right);
        }
        res.push(cn.val);
    }
    return res;
}

function dfsSolution2nd(cn,level,res){
    if(level >= res.length){
        res.push(cn.val);
    }
    if (cn.right) dfsSolution2nd(cn.right, level + 1,res);
    if (cn.left) dfsSolution2nd(cn.left, level + 1,res);
    return res;
}

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        finalRes = []
        queue = deque([root])
        
        def bfs(root):

            while queue:
                clLen = len(queue)
                for i in range(clLen):
                    cn = queue.popleft()
                    if cn.left:
                        queue.append(cn.left)
                    if cn.right:
                        queue.append(cn.right)

                    if i == clLen - 1:
                        finalRes.append(cn.val)
        def dfs(root):
            res = []
            def dfsSol(node, depth):
                if not node:
                    return

                
                if len(res) == depth:
                    res.append(node.val)

                dfsSol(node.right, depth + 1)
                dfsSol(node.left, depth + 1)
            
            dfsSol(root, 0)
            return res


        return dfs(root)        
        # bfs(root)

        
        # return finalRes
            
